### PR TITLE
RavenDB-18007 Get collection stats (shard wide)

### PR DESCRIPTION
**Pull request by efratshenhar**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18007

### Additional description

New endpoints for getting collection stats in sharded database

_Please delete below the options that are not relevant_

### Type of change

- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
 
### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    {
+
+        [RavenShardedAction("/databases/*/collections/stats", "GET")]
+        public async Task GetCollectionStats()
+        {
+            var stats = new CollectionStatistics();
+            var tasks = new List<Task>();
+            var res = new List<RavenCommand<CollectionStatistics>>();
+
+            for (int i = 0; i < ShardedContext.ShardCount; i++)
```
Comment (ml054) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs]: do we want to introduce some abstraction for that in parent method? I assume that scatter-collect will be commonly used pattern when dealing with sharded database group.

Comment (ayende) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs]: Not a bad idea, no. We typically want to send a command to all shards and then wait for it.

File: src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs [hunk 2]
```diff
@@ -76,5 +79,21 @@ public override async Task WaitForIndexToBeApplied(TransactionOperationContext c
...
+        {
+            var tasks = new List<Task>();
+
+            for (int i = 0; i < ShardedContext.ShardCount; i++)
+            {
+                var co = ContextPool.AllocateOperationContext(out JsonOperationContext context);
+                res.Add(commands[i].GetCommand(ShardedContext.RequestExecutors[i].Conventions, context));
+                var task = ShardedContext.RequestExecutors[i].ExecuteAsync(res[i], context);
+                task.ContinueWith(_ => co.Dispose());
+                tasks.Add(task);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs]: you probably want to add here result of continuation, not the original task?

File: src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace Raven.Server.Documents.ShardedHandlers.ContinuationTokens
+{
+    public abstract class ContinuationToken : IDynamicJson
+    {
+        public const string ContinuationTokenQueryString = "continuation-token";
+        public const string PropertyName = "ContinuationToken";
+        private static ConcurrentDictionary<Type, Func<BlittableJsonReaderObject, ContinuationToken>> ConverterCache =
+            new ConcurrentDictionary<Type, Func<BlittableJsonReaderObject, ContinuationToken>>();
+        public string ToBase64(JsonOperationContext context)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs:19]: to async

File: src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            var json = context.ReadObject(ToJson(), $"token for {GetType()}");
+            return Convert.ToBase64String(Encoding.UTF8.GetBytes(json.ToString()));
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs]: use .ToStringAsync

File: src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Documents.ShardedHandlers.ContinuationTokens
+{
+    public abstract class ContinuationToken : IDynamicJson
+    {
+        public const string ContinuationTokenQueryString = "continuation-token";
+        public const string PropertyName = "ContinuationToken";
+        private static ConcurrentDictionary<Type, Func<BlittableJsonReaderObject, ContinuationToken>> ConverterCache =
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs]: use Dictionary and same technique as we do in CachingJsonConverter

File: src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                var results = (ExtendedStreamResult)(await ShardExecutor.ExecuteParallelForAllAsync(op));
+
+                long numberOfResults;
+                long totalDocumentsSizeInBytes;
+                using (var cancelToken = CreateOperationToken())
+                await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream()))
+                {
+                    var documents = GetDocuments(results, token, cancelToken.Token);
+                    writer.WriteStartObject();
+                    writer.WritePropertyName("Results");
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs]: nameof

File: src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                    (numberOfResults, totalDocumentsSizeInBytes) = await writer.WriteDocumentsAsync(context, documents, metadataOnly: false, cancelToken.Token);
+                    writer.WriteComma();
+                    writer.WritePropertyName(ContinuationToken.PropertyName);
+                    writer.WriteString(token.ToBase64(context));
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs]: this needs to end up as an extension method

File: src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                               pagingContinuation, token))
+            {
+                yield return result.Item;
+            }
+        }
+
+        private ShardStreamItem<Document> BlittableToStreamDocument(BlittableJsonReaderObject json)
+        {
+            var metadata = json.GetMetadata();
+            metadata.TryGet(Constants.Documents.Metadata.Id, out LazyStringValue id);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs]: use GetId extension method?

File: src/Raven.Server/Documents/Sharding/ShardedContext.Streaming.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+            public int Start;
+
+            public DynamicJsonValue ToJson()
+            {
+                return new DynamicJsonValue { [nameof(Start)] = Start, [nameof(Shard)] = Shard };
+            }
+        }
+    }
+
+    public class ExtendedStreamResult : StreamResult
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedContext.Streaming.cs]: Combined?

File: src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs [hunk 10]
```diff
@@ -82,5 +84,21 @@ public override async Task WaitForIndexToBeApplied(TransactionOperationContext c
                 await tasks.WhenAll();
             }
         }
+
+        internal async Task GetShardsResults<T>(List<RavenCommand<T>> res, List<IMaintenanceOperation<T>> commands)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs]: remove this, not used

File: src/Voron/Data/Fixed/FixedSizeIterators.cs [hunk 11]
```diff
@@ -164,6 +164,9 @@ public bool Skip(long count)
                 if (count != 0)
                     _pos += count;
 
+                if (_pos < 0)
```
Comment (ppekrol) [inline src/Voron/Data/Fixed/FixedSizeIterators.cs:167]: cc @arekpalinski maybe we need to move this to 4.2?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs (hunk 7 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs]: done

File: src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedCollectionHandler.cs]: done

File: src/Raven.Server/Documents/Sharding/ShardedContext.Streaming.cs (hunk 9 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/ShardedContext.Streaming.cs]: done

File: src/Voron/Data/Fixed/FixedSizeIterators.cs (hunk 11 above)
Comment (arekpalinski) [reply src/Voron/Data/Fixed/FixedSizeIterators.cs:167]: @efratshenhar how did you run into that? Are you able to write a low level Voron test for that?

In general, yes - we should apply it to 4.2.

File: src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ContinuationToken.cs]: done

File: src/Voron/Data/Fixed/FixedSizeIterators.cs (hunk 11 above)
Comment (karmeli87) [reply src/Voron/Data/Fixed/FixedSizeIterators.cs:167]: I use the streaming... and if I put `start` param which is greater then the total entries we will end up with a negative value here which will cause an AVE :)

Comment (karmeli87) [reply src/Voron/Data/Fixed/FixedSizeIterators.cs:167]: https://issues.hibernatingrhinos.com/issue/RavenDB-18079

**Pull request merged**